### PR TITLE
Allow use of --copy-ec with --from-pr

### LINKED ISSUE #3178: eb --copy-ec does not work with --from-pr yet
would be nice to have that.

### PR DESCRIPTION
**Pull request by ocaisa**
edit: fixes #3178

### CODE REVIEW

**Comment by boegel** [comment]
@ocaisa Are you up for enhancing the test for `--copy-ec` to cover the use case (combo with `--from-pr`) you're fixing here?

**Comment by ocaisa** [comment]
I'm working on it now

**Comment by ocaisa** [comment]
@boegel I need a hint here, using `--from-pr` seems to require all the GitHub setup

**Comment by ocaisa** [comment]
This falls short, it won't also copy over the required patch files (having said that, the option `--copy-ec` never promised that)

**Comment by ocaisa** [comment]
@boegel Ready for review

File: test/framework/options.py [hunk 1]
```diff
@@ -1072,6 +1072,20 @@ def check_copied_files():
         self.assertTrue(os.path.exists(copied_toy_cwd))
         self.assertEqual(read_file(copied_toy_cwd), toy_ec_txt)
 
+        # Test --copy-ec coupled with --from-pr
+        all_ecs_pr8007 = [
```
[the remark this answers is not in the document]
Comment (boegel) [reply test/framework/options.py:1076]: @ocaisa Please fix this minor issue first, so we have a green light in the tests?

File: easybuild/main.py [hunk 2]
```diff
@@ -357,6 +373,12 @@ def main(args=None, logfile=None, do_build=None, testing=False, modtool=None):
                     log=_log, opt_parser=eb_go.parser, exit_on_error=not testing)
     _log.debug("Paths: %s", paths)
 
+    # if we are only copying ec's for a specific PR we can do that now
+    if options.copy_ec and options.from_pr:
+        determined_paths = [path[0] for path in paths]
+        copy_ecs_to_target(determined_paths, target_path)
```
Comment (ocaisa) [inline easybuild/main.py:379]: The change to use an absolute path for `target_path` was required because at this point, if we use `--from-pr`, we are inside a temporary directory so a relative path points to the wrong place.

File: easybuild/main.py [hunk 3]
```diff
@@ -357,6 +373,12 @@ def main(args=None, logfile=None, do_build=None, testing=False, modtool=None):
                     log=_log, opt_parser=eb_go.parser, exit_on_error=not testing)
     _log.debug("Paths: %s", paths)
 
+    # if we are only copying ec's for a specific PR we can do that now
+    if options.copy_ec and options.from_pr:
+        determined_paths = [path[0] for path in paths]
```
Comment (boegel) [inline easybuild/main.py:378]: @ocaisa This looks a bit strange, but the entries in `paths` are basically tuples of `(<path>, <bool>)`, where the `<bool>` indicates whether the path is generated or not.

I know you copy-pasted this from parts below, but this is probably clearer:

```python
determined_paths = [path for (path, _) in paths]
copy_ecs_to_target(determined_paths, target_path)
```

or just

```python
copy_ecs_to_target([path for (path, _) in paths], target_path)
```

or event


```python
copy_ecs_to_target([x for (x, _) in paths], target_path)
```

Please change it below too, to keep things consistent.

File: easybuild/main.py [hunk 4]
```diff
@@ -176,6 +176,24 @@ def run_contrib_style_checks(ecs, check_contrib, check_style):
     return check_contrib or check_style
 
 
+def copy_ecs_to_target(determined_paths, target_path, prefix=False):
```
Comment (boegel) [inline easybuild/main.py:179]: Doesn't this belong in `easybuild.framework.easyconfig.tools`?

Ideally we don't add much stuff to `main.py` (since then it can only be used here, nowhere else)

File: easybuild/main.py [hunk 5]
```diff
@@ -321,17 +343,11 @@ def main(args=None, logfile=None, do_build=None, testing=False, modtool=None):
     # determine paths to easyconfigs
     determined_paths = det_easyconfig_paths(categorized_paths['easyconfigs'])
 
-    if (options.copy_ec and not tweaked_ecs_paths) or options.fix_deprecated_easyconfigs or options.show_ec:
+    if (options.copy_ec and not (tweaked_ecs_paths or options.from_pr)) or options.fix_deprecated_easyconfigs or \
+            options.show_ec:
```
Comment (boegel) [inline easybuild/main.py:347]: We should clean this up a bit... How about this:

```python
# only copy easyconfigs here if we're not using --try-* or --from-pr (that's are handled below)
copy_ec = options.copy_ec and not (tweaked_ecs_paths or options.from_pr)
if copy_ec or options.fix_deprecated_easyconfigs or options.show_ec:

File: easybuild/main.py [hunk 6]
```diff
@@ -357,6 +373,12 @@ def main(args=None, logfile=None, do_build=None, testing=False, modtool=None):
                     log=_log, opt_parser=eb_go.parser, exit_on_error=not testing)
     _log.debug("Paths: %s", paths)
 
+    # if we are only copying ec's for a specific PR we can do that now
```
Comment (boegel) [inline easybuild/main.py:376]: `for a specific PR` -> `from a specific PR`

File: test/framework/options.py [hunk 7]
```diff
@@ -1072,6 +1072,49 @@ def check_copied_files():
...
+        stdout = mocked_main(args)
+        self.assertEqual(stdout, '2 file(s) copied to %s' % test_working_dir)
+        # check that the two easyconfigs exist
+        self.assertTrue(os.path.exists(os.path.join(test_working_dir, all_ecs_pr8007[0])))
+        remove_file(os.path.join(test_working_dir, all_ecs_pr8007[0]))
+        self.assertTrue(os.path.exists(os.path.join(test_working_dir, all_ecs_pr8007[2])))
+        remove_file(os.path.join(test_working_dir, all_ecs_pr8007[2]))
+        # ...but the patch doesn't
+        self.assertFalse(os.path.exists(os.path.join(test_working_dir, all_ecs_pr8007[1])))
+        remove_file(os.path.join(test_working_dir, all_ecs_pr8007[1]))
```
Comment (boegel) [inline test/framework/options.py:1096]: Hmm, would it be difficult to fix this? I would expect that the patch gets copied too...

We should also test copying a single easyconfig from a PR where multiple are touched?
Like `eb --copy-ec --from-pr 8007 Arrow-0.7.1-intel-2017b-Python-3.6.3.eb <target_dir>`

File: easybuild/main.py [hunk 8]
```diff
@@ -303,12 +321,16 @@ def main(args=None, logfile=None, do_build=None, testing=False, modtool=None):
             eb_file = find_easybuild_easyconfig()
             orig_paths.append(eb_file)
 
-    if len(orig_paths) == 1:
+    if not orig_paths:
+        # if no easyconfig files are specified and we are using --from-pr, use current directory as target directory
+        target_path = os.getcwd() if (options.copy_ec and options.from_pr) else None
```
Comment (boegel) [inline easybuild/main.py:326]: How about:

```python
# if only one easyconfig is specified, or if none are specified and we are using --from-pr,
# use current directory as target directory
if len(orig_paths) == 1 or (options.copy_ec and options.from_pr and len(orig_paths) == 0):
    target_path = os.getcwd()
elif orig_paths:
    ...

File: easybuild/main.py [hunk 9]
```diff
@@ -303,12 +321,16 @@ def main(args=None, logfile=None, do_build=None, testing=False, modtool=None):
...
+    elif len(orig_paths) == 1 and not (options.copy_ec and options.from_pr):
         # if only one easyconfig file is specified, use current directory as target directory
         target_path = os.getcwd()
     elif orig_paths:
         # last path is target when --copy-ec is used, so remove that from the list
-        target_path = orig_paths.pop() if options.copy_ec else None
+        # use the absolute path as some use cases occur in a temporary directory
```
Comment (boegel) [inline easybuild/main.py:332]: In which case are we copying to the target path from a different directory?
I'm guessing that's with `--from-pr` (we should mention it here rather than leave ourselves guessing in the future)

Perhaps that should be considered a bug?

File: test/framework/options.py [hunk 10]
```diff
@@ -2067,15 +2110,15 @@ def test_try_with_copy(self):
 
         self.mock_stdout(True)
         self.mock_stderr(True)
-        tweaked_ecs_dir = os.path.join(self.test_buildpath, 'my_tweaked_ecs')
-        self.eb_main(args + ['--try-software=foo,1.2.3', '--try-toolchain=gompi,2018a', tweaked_ecs_dir],
-                     verbose=True, raise_error=True)
+        tweaked_ecs_dir = os.path.join(self.test_buildpath)
+        os.chdir(tweaked_ecs_dir)
```
Comment (boegel) [inline test/framework/options.py:2114]: Why was this changed? Doesn't this signal a (breaking) change in behavior?

**Review by boegel — CHANGES_REQUESTED** [review]

File: test/framework/options.py (hunk 7 above)
Comment (ocaisa) [reply test/framework/options.py:1096]: I can easily copy *all* patches that are alongside an easyconfig, but I don't know how to (easily) filter the ones that are part of a PR

File: easybuild/main.py (hunk 3 above)
Comment (ocaisa) [reply easybuild/main.py:378]: Hmm, not sure what you mean by doing it below, the case there is very different

File: easybuild/main.py (hunk 2 above)
Comment (ocaisa) [reply easybuild/main.py:379]: I agree that this behaviour is probably a bit of a bug

File: test/framework/options.py (hunk 10 above)
Comment (ocaisa) [reply test/framework/options.py:2114]: Yeah, it does change behaviour but it's not a breaking change. For the case where you were tweaking things, it always copied the tweaked files to a new directory (even if there was only one path in the list, which is the case in the test). This is different behaviour to what happens in the original use case, where if there was only one file, it would do a simple copy to the destination path. I thought it was probably more important to make the behaviour consistent in all cases.

File: test/framework/options.py (hunk 7 above)
Comment (ocaisa) [reply test/framework/options.py:1096]: Regarding the additional test, it is not going to pick that up, according to the existing logic, it is going to assume that the additional argument is the target path. I'm not sure what logic I could use to get around that, I'd need to know if the filename(s) appear(s) in the PR.

Comment (ocaisa) [reply test/framework/options.py:1096]: Maybe I can do it before I copy the files, if the basename of the target path matches one of the paths then I assume that is what was meant and only copy that single file to the basedir of the target path

File: easybuild/main.py (hunk 3 above)
Comment (boegel) [reply easybuild/main.py]: `regtest([path[0] for path in paths] or ...` -> `regtest([x for (x, _) in paths] or`

File: test/framework/options.py (hunk 10 above)
Comment (ocaisa) [reply test/framework/options.py:2114]: I can revert this, but I should probably take the same approach with `--from-pr`, where you don't necessarily know how many ecs you will end up with.

**Comment by ocaisa** [comment]
I think I've made this overly complicated, now that I've had a look at existing GitHub functionality I could have just used `fetch_files_from_pr()` and this would be done

**Review by boegel — CHANGES_REQUESTED** [review]
I spent quite a bit of time cleaning things up on top of these changes to clean things up a bit, see #3482
